examples/ecs/computed_states.rs

```diff
@@ -0,0 +1,209 @@
+//! This example illustrates how to use [`ComputedStates`] for more complex state transitions.
+//!
+//! In this case, we're transitioning from a `Menu` state to an `InGame` state, with the addition
+//! of the capacity to pause and unpause if we are in game using the ESCAPE button on the keyboard.
+
+use bevy::prelude::*;
+
+fn main() {
+    // We start the setup like we did in the states example.
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .init_state::<AppState>()
+        .add_systems(Startup, setup)
+        .add_systems(OnEnter(AppState::Menu), setup_menu)
+        .add_systems(Update, menu.run_if(in_state(AppState::Menu)))
+        .add_systems(OnExit(AppState::Menu), cleanup_menu)
+        // We only want to run the [`setup_game`] function when we enter the [`AppState::InGame`] state, regardless
+        // of whether the game is paused or not. So we add an [`InGame`] derived state, and rely on that.
+        .add_computed_state::<InGame>()
+        .add_systems(OnEnter(InGame), setup_game)
+        // We want the color change and the toggle_pause systems to ignore the paused condition, so we can use the [`InGame`] derived
+        // state here as well.
+        .add_systems(
+            Update,
+            (toggle_pause, change_color).run_if(in_state(InGame)),
+        )
+        // However, we only want the movement to happen if we are not in a paused state, so here we
+        // use the full AppState: `AppState::InGame { paused: false }`.
+        .add_systems(
+            Update,
+            movement.run_if(in_state(AppState::InGame { paused: false })),
+        )
+        .add_systems(Update, log_transitions)
+        .run();
+}
+
+#[derive(Debug, Clone, Copy, Default, Eq, PartialEq, Hash, States)]
+enum AppState {
+    #[default]
+    Menu,
+    InGame {
+        paused: bool,
+    },
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, States)]
+struct InGame;
```

Comment:
As mentioned earlier, I wouldn't use a ZST as my first choice, rather an enum.

This may be bias from one of the prototypes I wrote, but I'd probably use the same enum, but without any sub-fields here:
```rs
// sub-state-free
enum SsfAppState {
    #[default]
    Menu,
    InGame,
}
```
(which btw would be amazing if we had a macro for making those)

It's mostly the state-removal API that's I'm unsure about though, I'll have to think about it for a bit.

Reply:
The ZST support is a side effect of being able to use `Option`. See above for my reasons in support of the state removal API - but I am also including a standard `enum` state in the updated example! (it's in the works, just had a busy day so wasn't able to finish it)